ssl/quic/quic_impl.c

```diff
@@ -4623,6 +4632,274 @@ static QUIC_CONNECTION *create_qc_from_incoming_conn(QUIC_LISTENER *ql, QUIC_CHA
+                                             size_t token_len)
+{
+    QUIC_TOKEN *new_token;
+    size_t hashkey_len = 0;
+    size_t addr_len = 0;
+    int family;
+    unsigned short port;
+    int *famptr;
+    unsigned short *portptr;
+    uint8_t *addrptr;
+
+    if (!BIO_ADDR_rawaddress(peer, NULL, &addr_len))
+        return NULL;
+    family = BIO_ADDR_family(peer);
+    port = BIO_ADDR_rawport(peer);
+
+    hashkey_len += sizeof(int); /* hashkey(family) */
+    hashkey_len += sizeof(unsigned short); /* hashkey(port) */
+    hashkey_len += addr_len; /* hashkey(address) */
+
+    new_token = OPENSSL_zalloc(sizeof(QUIC_TOKEN) + hashkey_len + token_len);
+    if (new_token == NULL)
+        return NULL;
+
+    if (!CRYPTO_NEW_REF(&new_token->references, 1)) {
+        OPENSSL_free(new_token);
+        return NULL;
+    }
+
+    new_token->hashkey_len = hashkey_len;
+    /* hashkey is allocated inline, immediately after the QUIC_TOKEN struct */
+    new_token->hashkey = (uint8_t *)(new_token + 1);
+    /* token buffer follows the hashkey in the inline allocation */
+    new_token->token = new_token->hashkey + hashkey_len;
+    new_token->token_len = token_len;
+    famptr = (int *)new_token->hashkey;
+    portptr = (unsigned short *)(famptr + 1);
+    addrptr = (uint8_t *)(portptr + 1);
+    *famptr = family;
+    *portptr = port;
+    if (!BIO_ADDR_rawaddress(peer, addrptr, NULL)) {
+        ossl_quic_free_peer_token((QTOK *)new_token);
+        return NULL;
+    }
+    if (token != NULL)
+        memcpy(new_token->token, token, token_len);
+    return new_token;
+}
+
+int ossl_quic_set_peer_token(SSL_CTX *ctx, BIO_ADDR *peer,
+                             const uint8_t *token, size_t token_len)
+{
+    SSL_TOKEN_STORE *c = ctx->tokencache;
+    QUIC_TOKEN *tok, *old = NULL;
+
+    if (ctx->tokencache == NULL)
+        return 0;
+
+    tok = ossl_quic_build_new_token(peer, (uint8_t *)token, token_len);
+    if (tok == NULL)
+        return 0;
+
+    /* we might be sharing this cache, lock it */
+    ossl_crypto_mutex_lock(c->mutex);
+
+    old = lh_QUIC_TOKEN_retrieve(c->cache, tok);
+    if (old != NULL) {
+        lh_QUIC_TOKEN_delete(c->cache, old);
+        ossl_quic_free_peer_token((QTOK *)old);
+    }
+    lh_QUIC_TOKEN_insert(c->cache, tok);
+
+    ossl_crypto_mutex_unlock(c->mutex);
+    return 1;
+}
+
+int ossl_quic_get_peer_token(SSL_CTX *ctx, BIO_ADDR *peer,
+                             uint8_t **token, size_t *token_len,
+                             QTOK **token_free_ptr)
+{
+    SSL_TOKEN_STORE *c = ctx->tokencache;
+    QUIC_TOKEN *key = NULL;
+    QUIC_TOKEN *tok = NULL;
+    int ret;
+    int rc = 0;
+
+    if (c == NULL)
+        return 0;
+
+    key = ossl_quic_build_new_token(peer, NULL, 0);
+    ossl_crypto_mutex_lock(c->mutex);
+    tok = lh_QUIC_TOKEN_retrieve(c->cache, key);
+    if (tok != NULL) {
+        *token = tok->token;
+        *token_len = tok->token_len;
+        *token_free_ptr = (QTOK *)tok;
```

Comment:
I am confused why we have two different types for this. What is the difference between `QTOK` and `QUIC_TOKEN`?

Reply:
QUIC_VALIDATION_TOKEN is the that is used to create/parse and validate tokens on the wire.  It is private to quic_port.c

QUIC_TOKEN is the structure that is stored as the data element in the token store.  It hold the opaque, encrypted blob that is the token, along with refcounting data for user of the token.

QTOK is a typedef void representing the pointer to be freed when ossl_quic_get_peer_token returns a token buf.  It is an alias for QUIC_TOKEN.  Note the comments near ossl_quic_build_new_token explaining the token layout.

I can export QUIC_TOKEN to other files to eliminate the need for QTOK